Add script to report rules

AyanSinhaMahapatra opened the pull request:
Partially adresses #2566

<!--
Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!
* [x] - Checked Box
* [ ] - Unchecked Box
-->

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

<!--
We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged**
if your tests fail at first!
If tests do fail, click on the red `X` to learn why by reading the logs.
Thanks!
-->

<!-- Don't forget to Signoff -->

AyanSinhaMahapatra committed:
Add script to report rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented:
@mjherzog this is ready for you to look at.

Here's the help text for this script:

```
Usage: report_license_rules.py [OPTIONS]

  Write Licenses/Rules from scancode into a CSV file with all details. Output
  can be optionally filtered by category/license-key.

Options:
  -l, --licenses FILE       Write all Licenses data to the csv FILE.
  -r, --rules FILE          Write all Rules data to the csv FILE.
  -c, --category STRING     An optional filter to only output licenses/rules
                            of this category. Example STRING: `permissive`.
  -k, --license-key STRING  An optional filter to only output licenses/rules
                            which has this license key.Example STRING: `mit`.
  -t, --with-text           Also include the license/rules texts.Note that
                            this increases the file size significantly.
  -h, --help                Show this message and exit.

```

You have to checkout this branch in a configured virtualenv and run the script if you want to check this out now. 

Attaching sample ouput files:

[licence_data.csv](https://github.com/nexB/scancode-toolkit/files/7107763/lic_data.csv)
[rule_data.csv](https://github.com/nexB/scancode-toolkit/files/7107762/rule_data.csv)

mjherzog commented:
@AyanSinhaMahapatra Thank you for running the sample with:
python ./etc/scripts/licenses/report_license_rules.py -l comm_licenses_data.csv -r comm_rule_data.csv -t -c "Commercial"
It looks good overall, but there is strange data in the identifier column - it looks like some kind of "bleed" from license text. Easy to see the problem if you sort by identifier because there is a very big chunk at the beginning and the other examples interspersed among the "true" identifiers.

mjherzog commented:
I see clues for Accellera and Hacktivismo in the errant text.
If we have text too large for a single csv/Excel cell we could just truncate it. This tool is needed most for licenses with relatively sparse text for most rules like commercial-license.

pombredanne commented on etc/scripts/licenses/report_license_rules.py:
```diff
@@ -0,0 +1,198 @@
+
+
+LICENSES_FIELDNAMES = [
+    'key', 'short_name', 'name', 'category', 'owner',  'text', 'words_count', 'notes',
```
text is always going to be problematic... using the first `[:200]` characters is going to be enough

pombredanne commented on etc/scripts/licenses/report_license_rules.py:
```diff
@@ -0,0 +1,198 @@
+    'license_filename', 'faq_url', 'ignorable_authors',
+    'ignorable_copyrights', 'ignorable_holders', 'ignorable_urls', 'ignorable_emails',
+    'osi_license_key', 'osi_url', 'other_spdx_license_keys',
+]
```
Add the licensedb URL too

pombredanne commented on etc/scripts/licenses/report_license_rules.py:
```diff
@@ -0,0 +1,198 @@
+
+def write_data_to_csv(data, output_csv, fieldnames):
+    
+    with open(output_csv,'w',encoding='utf-8-sig',newline='') as f:
```
`,encoding='utf-8-sig',newline=''` are suspicious to me... IMHO you are best of without them

pombredanne commented on etc/scripts/licenses/report_license_rules.py:
```diff
@@ -0,0 +1,198 @@
+            w.writerow(entry)
+
+
+def filter_by_attribute(data, attribute, expected):
```
may be required_key is a better name than expected?

pombredanne commented on etc/scripts/licenses/report_license_rules.py:
```diff
@@ -0,0 +1,198 @@
+
+def filter_by_attribute(data, attribute, expected):
+    
+    filtered_output = []
```
my be this is readable too?

```Python
return [entry for entry in data if entry.get(attribute) == expected]
```

pombredanne commented on etc/scripts/licenses/report_license_rules.py:
```diff
@@ -0,0 +1,198 @@
+def flatten_output(data):
+
+    if not isinstance(data, list):
+        return
```
say I have a tuple or a dict?... the flatten_output is then empty? The implicit exclusion of non-list is a bit surprising. If you only ever expect a list, then instead use `assert isinstance(data, list)`

pombredanne commented on etc/scripts/licenses/report_license_rules.py:
```diff
@@ -0,0 +1,198 @@
+
+    output = []
+    for entry in data:
+        if not isinstance(entry, dict):
```
You should assert instead

pombredanne reviewed (approved):
See a few nits for your consideration!
Feel free to merge.

AyanSinhaMahapatra committed:
Add requested changes to script

Add licensedb url, modify csv output to correct bugs, only add first 200 characters
in case of text, modify functions and add docstrings.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented:
@pombredanne added requested changes.

@mjherzog added some changes that might fix your problems earlier. Made texts have only first 200 characters and removed some probably problematic csv file handling. Adding new csv file reports, do check if the bugs are resolved.

The commands used to generate reports:
```
python ./etc/scripts/licenses/report_license_rules.py -l licenses_data.csv -r rule_data.csv -t
python ./etc/scripts/licenses/report_license_rules.py -l comm_licenses_data.csv -r comm_rule_data.csv -t -c "Commercial"
```

Reports:
[licenses_data.csv](https://github.com/nexB/scancode-toolkit/files/7115780/licenses_data.csv)
[rule_data.csv](https://github.com/nexB/scancode-toolkit/files/7115782/rule_data.csv)
[comm_licenses_data.csv](https://github.com/nexB/scancode-toolkit/files/7115785/comm_licenses_data.csv)
[comm_rule_data.csv](https://github.com/nexB/scancode-toolkit/files/7115789/comm_rule_data.csv)
